rfc/0003-policy-signing.md

```diff
@@ -0,0 +1,440 @@
+  - john@acme.org
+annotations:
+  env: prod
+  foo: bar
+```
+
+The file is composed by three sections, which we will analyze in the next
+sections.
+
+
+### Public keys
+
+This is a list of the public keys of the authors that are trusted:
+
+```yaml
+public_keys:
+  threshold: 2
+  keys:
+    alice: |
+      -----BEGIN PUBLIC KEY-----
+      MBFKHFDGHKIJH0CAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbDE6HJSGAVD
+      BVDF6SKFSF87AASUspkQsN3FO4iyWodCy5j3o0CdIJD/KJHDJFHDFIu6sA==
+      -----END PUBLIC KEY-----
+    bob: |
+      -----BEGIN PUBLIC KEY-----
+      AUYSDHKJFGSHGSHCAQYIKoZIzj0DAQcDQgAEX0HFTtCfTtPmkx5p1RbtwDE1EVzu
+      wjQs1cCRKb5Pz/yUspkQsN3FO4iyWodCy5j3o0CdIJD/1gvq98pf4IG9tA==
+      -----END PUBLIC KEY-----
+    john:|
+      -----BEGIN PUBLIC KEY-----
+      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAELKhD7F5OKy77Z582Y6h0u1J3GNA+
+      kvUsh4eKpd1lwkDAzfFDs7yXEExsEkPPuiQJBelDT68n7PDIWB/QEY7mrA==
+      -----END PUBLIC KEY-----
+```
+
+* `keys`: list of keys that are trusted. The owner's name is not relevant during
+  the verification process.
+* `threshold`: an integer number of keys whose signatures are required in order
+  to consider a policy as being properly signed.
+
+Given the above example, there are three keys defined and the threshold is set to
+`2`. That means all the policies loaded must be signed by two or more of these
+keys.
+
+This gives some resiliency if one of these keys is compromised.
+
+### Issuers
+
+It allows to define, on a per issuer basis, who are the trusted`Subject`s:
+
+```yaml
+issuers:
+  "https://token.actions.githubusercontent.com":
+  - https://github.com/flavio/
+  - https://github.com/kubewarden/
+  - https://github.com/alice/fancy-policy/
+  "https://github.com/login/oauth":
+  - john@acme.org
+```
+
+The string entered by the user is then used to match the beginning of the `Subject`
+found inside of a signature.
+
+For example, given the configuration from above:
+
+Subject | Verified | Reason |
+------------------|----------|--------|
+`https://github.com/flavio/policy-secure-pod-images/.github/workflows/release.yml@refs/heads/main` | ✅ | The policy is coming from `https://github.com/flavio/`
+`https://github.com/flavio/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.0.2` | ✅ | The policy is coming from `https://github.com/flavio/`
+`https://github.com/kubewarden/privileged-pods/.github/workflows/release.yml@refs/tags/v0.2.0` | ✅ | The policy is coming from `https://github.com/kubewarden/`
+`https://github.com/kubewarden/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.3.0` | ✅ | The policy is coming from `https://github.com/kubewarden/`
+`https://github.com/alice/fancy-policy/.github/workflows/cd.yml@refs/heads/main` | ✅ | The policy is coming from `https://github.com/alice/fancy-policy/`
+`https://github.com/alice/another-policy/.github/workflows/cd.yml@refs/heads/main` | ❌ | The policy is coming from a repository owned by Alice, but not trusted by the configuration
+
```

Comment:
I don't like to perform this kind of verification, I think it can lead to unexpected results.

For example, given the following configuration:

```yaml
issuers:
  "https://token.actions.githubusercontent.com":
  - https://github.com/flavio
```

Would lead to **trust** these kind of issuers:

* `https://github.com/flavio/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.0.2`
* `https://github.com/flavio-the-hacker/psp-capabilities/.github/workflows/release.yml@refs/tags/v0.0.2`

A missing `/` inside of the setting could lead to this unexpected behaviour.

Moreover, I might be paranoid, but I think someone might come up asking for a way to provide RegExp to validate the `Subject`. I think this would be super scary from a security POV, many things could go wrong that way.

I think it would be better to provide a more strict and explicit configuration:

```yaml
issuers:
  GitHubActions:
  - owner: flavio
    repositories:
      - repo1
      - repo2
   - owner: kubewarden
```

In that case, we would trust the `Subject`s generated by:
* user `flavio`: only `repo1` and `repo2`. All the other flavio's repos would not be trusted
* org `kubewarden`: all the repos are trusted

Pros:
  * more explicit and secure
  * less chances to shoot yourself in the foot

Cons:
  * Less generic
  * Could force us to provide more ad-hoc code for other environments (GitLab, Tekton, ...)

Reply:
I think that if we parse the list of subjects as URL's when they are not e-mail addresses this should not be a problem.

This being said: we can support both if we want. In a model in which we allow different keys as `issuers` children that might behave different:

```yaml
issuers:
  GitHubActions:
  - owner: flavio
    repositories:
      - repo1
      - repo2
  - owner: kubewarden
```

And also:

```yaml
issuers:
  GenericIssuers:
    "https://token.actions.githubusercontent.com":
      - https://github.com/flavio/repo1
      - https://github.com/flavio/repo2
      - https://github.com/kubewarden
```

Would be equivalent.

With this you have the best of both worlds I think. However, I doubt we really need to start with this from the early beginning. If not, then the question would be: should we start with `GenericIssuers` or with `GitHubActions` only? In my opinion, I would prefer to start with `GenericIssuers` only or with both.

I don't like that we have a single way of specifying issuers and that is tied with GitHub without an alternative.

Reply:
I also favor the approach of just having a generic issuer for now, and look for either url or email. I'm not sure on which yaml would be better UX so that we look for either url or email. Either having a `genericURLIssuer`, `genericEmailIssuer` or the like, or have tuples `issuer, email`, `issuer, url`.

Reply:
I would prefer to the keep what is described into the RFC, and just add at the end of the urls that users are providing.

To clarify:

```yaml
issuers:
  "https://token.actions.githubusercontent.com":
  - https://github.com/flavio/
  - https://github.com/kubewarden # -> this becomes https://github.com/kubewarden/
  "https://github.com/login/oauth":
  - john@acme.org
```

Reply:
Now obsolete, because of the changes done to `verification.yml` format